### PR TITLE
Update dumpsoftwareversions module

### PR DESCRIPTION
**Pull request by grst**
Highlights
 * extracts module name from fully qualified process name. If our assumption that modules with the same name from different subworkflows share the same versions is violated, an AssertionError is raised. 
 * swiched to the brand-new [module templates feature](https://www.nextflow.io/docs/latest/dsl2.html#module-templates). This allows having a separate python script in the modules directory :tada: 
 * Updated tests to check that it really works with two subworkflows.

### CODE REVIEW

File: tests/modules/custom/dumpsoftwareversions/main.nf [hunk 1]
```diff
@@ -6,19 +6,48 @@ include { FASTQC                      } from '../../../../modules/fastqc/main.nf
 include { MULTIQC                     } from '../../../../modules/multiqc/main.nf'
 include { CUSTOM_DUMPSOFTWAREVERSIONS } from '../../../../modules/custom/dumpsoftwareversions/main.nf'
 
+
+workflow fastqc1 {
+    take:
+    input
+
+    main:
+    FASTQC( input )
```
Comment (drpatelh) [inline tests/modules/custom/dumpsoftwareversions/main.nf:15]:
```suggestion
    FASTQC ( input )
```

File: tests/modules/custom/dumpsoftwareversions/main.nf [hunk 2]
```diff
@@ -6,19 +6,48 @@ include { FASTQC                      } from '../../../../modules/fastqc/main.nf
...
+    emit:
+    versions = FASTQC.out.versions
+}
+
+workflow fastqc2 {
+    take:
+    input
+
+    main:
+    FASTQC( input )
```
Comment (drpatelh) [inline tests/modules/custom/dumpsoftwareversions/main.nf:26]:
```suggestion
    FASTQC ( input )
```

File: tests/modules/custom/dumpsoftwareversions/main.nf [hunk 3]
```diff
@@ -6,19 +6,48 @@ include { FASTQC                      } from '../../../../modules/fastqc/main.nf
...
         [ file(params.test_data['sarscov2']['illumina']['test_1_fastq_gz'], checkIfExists: true),
         file(params.test_data['sarscov2']['illumina']['test_2_fastq_gz'], checkIfExists: true) ]
     ]
 
-    FASTQC  ( input )
-    MULTIQC ( FASTQC.out.zip.collect { it[1] } )
+    // Using subworkflows to ensure that the script can properly handle
+    // cases where subworkflows have a module with the same name.
+    fastqc1( input )
+    fastqc2( input )
```
Comment (drpatelh) [inline tests/modules/custom/dumpsoftwareversions/main.nf:43]:
```suggestion
    fastqc1 ( input )
    fastqc2 ( input )
```

File: modules/custom/dumpsoftwareversions/templates/dumpsoftwareversions.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    "yaml": yaml.__version__,
+}
+
+with open("$versions") as f:
+    versions_by_process = yaml.load(f, Loader=yaml.BaseLoader) | versions_this_module
+
+# aggregate versions by the module name (derived from fully-qualified process name)
+versions_by_module = {}
+for process, process_versions in versions_by_process.items():
+    module = process.split(":")[-1]
```
Comment (grst) [inline modules/custom/dumpsoftwareversions/templates/dumpsoftwareversions.py:59]: Or should this be
```python
process.split(":")[-1].split("_")[-1]
```

?

Comment (drpatelh) [reply modules/custom/dumpsoftwareversions/templates/dumpsoftwareversions.py:59]: What is the difference in terms of output?

File: modules/custom/dumpsoftwareversions/main.nf [hunk 5]
```diff
@@ -17,80 +17,5 @@ process CUSTOM_DUMPSOFTWAREVERSIONS {
...
-
-    with open("software_versions.yml", 'w') as f:
-        yaml.dump(workflow_versions, f, default_flow_style=False)
-    with open("software_versions_mqc.yml", 'w') as f:
-        yaml.dump(versions_mqc, f, default_flow_style=False)
-
-    with open('versions.yml', 'w') as f:
-        yaml.dump(module_versions, f, default_flow_style=False)
-    """
+    template 'dumpsoftwareversions.py'
```
Comment (JoseEspinosa) [inline modules/custom/dumpsoftwareversions/main.nf:20]: 😍

**Pull request merged**